lib/phoenix/router/helpers.ex

```diff
@@ -231,6 +240,36 @@ defmodule Phoenix.Router.Helpers do
     end
   end
 
+  def defhelper_catch_all({helper, routes}) do
+    valid_routes = Enum.map(routes, &(&1.opts))
+    # We are using -1 to avoid warnings in case a path has already been defined.
+
+    for suffix <- ["path", "url"] do
+      quote line: -1 do
+        def unquote(:"#{helper}_#{suffix}")(_conn_or_endpoint, action) do
+          Phoenix.Router.Helpers.raise_route_error(__MODULE__, "#{unquote(helper)}_#{unquote(suffix)}/2", action, unquote(valid_routes))
+        end
+
+        def unquote(:"#{helper}_#{suffix}")(_conn_or_endpoint, action, _params) do
+          Phoenix.Router.Helpers.raise_route_error(__MODULE__, "#{unquote(helper)}_#{unquote(suffix)}/3", action, unquote(valid_routes))
+        end
+      end
+    end
+  end
+
+  @doc false
+  def raise_route_error(module, helper, action, valid_routes) do
+    module_without_prefix = module |> to_string() |> String.split(".", parts: 2) |> Enum.at(1)
+    valid_actions = Enum.join(["" | valid_routes], "\n * :")
+
+    message = "no function clause matching in #{module_without_prefix}.#{helper}\n\n" <>
+              "This is likely due to the use of :#{action} as the function name.\n\n" <>
+              "Only the following atoms can be used:\n" <>
+              valid_actions
```

Comment:
@chrismccord I prefer the wording. I think displaying the arity of the function is important though.